scannerc.go

```diff
@@ -633,15 +633,20 @@ func yaml_parser_fetch_more_tokens(parser *yaml_parser_t) bool {
 			// Queue is empty.
 			need_more_tokens = true
 		} else {
+			// Check for any now stale keys in a required position.
+			if !yaml_parser_stale_simple_keys(parser) {
+				return false
+			}
 			// Check if any potential simple key may occupy the head position.
 			for i := len(parser.simple_keys) - 1; i >= 0; i-- {
 				simple_key := &parser.simple_keys[i]
+				if !simple_key.possible {
+					break
+				}
```

Comment:
Again, it's possible I misunderstand the algorithm since I didn't put the original logic in place here, but this doesn't look right to me. That's stopping the search if the most recently scanned depth level was marked as not a possible simple key, but that's not the actual question being asked by the logic. The question is: can the head token -- that is, the token that is about to be consumed by the parser -- be a simple key still, given everything else we parsed so far? We can't tell if the head token is possible or not without finding it, which is done below.

This will certainly speed things up because it's not trying to find the head anymore, but it doesn't seem right.

Am I missing something?

Reply:
I think you're right. I'm not sure what the negative test case would be, but I agree that the spirit of this piece of logic seems to not be maintained.

From what I can tell, this function's intent is to fetch enough tokens to determine whether the next value fetched for the parser is going to be an implicit simple-key (e.g. `[[[]]]: foo`). If can't be a simple key, and that was required, fail. If it can't be a simple key but we didn't require one, then let the parser consume the non-simple-key value.

If we want to keep the check for "can a simple key occupy the head position" as a simple scan, I think we always end up w/ potentially 1024 keys to scan, before we either decide we need more tokens or that we are ready to return w/ a token available (or if we scan from the bottom, up to 10,000 keys to scan).

If the only key we actually care about is the one that potentially starts with the token at the parser head, then we only really need to check that. So now I am back to my other proposal of maintaining a map of possible simple_keys :).

Reply:
> If the only key we actually care about is the one that potentially starts with the token at the parser head, then we only really need to check that. So now I am back to my other proposal of maintaining a map of possible simple_keys :).

Yes, your map proposal was never considered incorrect. It's a reasonable way to handle a very complex case. The reason why I've suggested resisting it is because as far as I can see this sort of pattern will only be found in unusual documents, typically abusive. Real documents will pretty much never get even remotely close to this being an issue.

That said, if that's what it takes, so be it, and the change on top of the new logic also looks simpler compared to the previous proposal we looked into.

I'll try to merge this today or tomorrow at a window that allows me to pay attention to any issues.